MiauAPI/Services/ProductService.cs

```diff
@@ -0,0 +1,209 @@
+            }
+            else
+            {
+                dbProducts = dbProducts
+                                .Where(p => ((p.Price * (1 - p.Discount)) >= productParameters.MinPrice)
+                                         && ((p.Price * (1 - p.Discount)) <= productParameters.MaxPrice))
+                                .ToList();
+            }
+
+            if (!CheckQuery(dbProducts, out descriptionError))
+            {
+                return new NotFoundObjectResult(new ErrorResponse(descriptionError));
+            }
+        }
+
+        // List products with an active discount
+        if (productParameters.ActiveDiscount)
+        {
+            dbProducts = dbProducts
+                            .Where(p => p.Discount > 0)
+                            .ToList();
+
+            if (!CheckQuery(dbProducts, out var descriptionError))
+            {
+                return new NotFoundObjectResult(new ErrorResponse(descriptionError));
+            }
+        }
+
+        // List products with specifics tags
+        if (productParameters.Tags != 0)
+        {
+            dbProducts = dbProducts
+                            .Where(p => p.Tags.HasFlag(productParameters.Tags))
+                            .ToList();
+
+            if (!CheckQuery(dbProducts, out var descriptionError))
+            {
+                return new NotFoundObjectResult(new ErrorResponse(descriptionError));
+            }
+        }
+
+        // Sort products by Price Asc (default), PriceDesc,  DiscountAsc, DiscountDesc
+        switch (productParameters.SortParameter)
+        {
+            case SortParameter.PriceAsc:
+                dbProducts = dbProducts
+                            .OrderBy(p => (p.Price * (1 - p.Discount)))
+                            .ToList();
+                break;
+            case SortParameter.PriceDesc:
+                dbProducts = dbProducts
+                            .OrderByDescending(p => (p.Price * (1 - p.Discount)))
+                            .ToList();
+                break;
+            case SortParameter.DiscountAsc:
+                dbProducts = dbProducts
+                            .OrderBy(p => p.Discount)
+                            .ToList();
+                break;
+            case SortParameter.DiscountDesc:
+                dbProducts = dbProducts
+                            .OrderByDescending(p => p.Discount)
+                            .ToList();
+                break;
+        }
+
+        dbProducts = dbProducts
+                            .OrderByDescending(p => p.IsActive)
+                            .ToList();
+
+        var dbProductsPaged = PagedList<ProductEntity>.ToPagedList(
+                        dbProducts,
+                        productParameters.PageNumber,
+                        productParameters.PageSize);
+
+        return new OkObjectResult(new GetProductResponse(dbProductsPaged));
+    }
```

Comment:
Esse método é grande demais e faz coisas demais. Subdividide-o em métodos menores.
Se você sentir que for necessário, crie endpoints específicos para cada tipo de busca.